File: packages/testkit/swift/Sources/BridgethingTestKit/WireDriver.swift
import BridgethingGateway
import BridgethingSchema
import Foundation

public enum WireDriverError: Error, Sendable {
    case timeout
    case decodeFailed
}

public actor WireDriver {
    private let adapter: InMemoryAdapter
    private let codec: Codec
    public let deviceId: String

    private var pending: [UUID: CheckedContinuation<GatewayToBridgeMsg, Error>] = [:]
    private var outbound: [GatewayToBridgeMsg] = []
    private var waiters: [(id: UUID, predicate: @Sendable (GatewayToBridgeMsg) -> Bool, cont: CheckedContinuation<GatewayToBridgeMsg, Error>)] = []
    private var pumpTask: Task<Void, Never>?

    public init(
        adapter: InMemoryAdapter,
        deviceId: String = "carthing-test",
        codec: Codec = Codec()
    ) {
        self.adapter = adapter
        self.deviceId = deviceId
        self.codec = codec
    }

    /// Begin draining outbound frames. Call once, after the companion is started.
    public func start() {
        guard pumpTask == nil else { return }
        let stream = adapter.sentFrames
        let codec = self.codec
        pumpTask = Task { [weak self] in
            for await (_, frame) in stream {
                guard let msg = try? codec.decode(GatewayToBridgeMsg.self, from: frame) else { continue }
                await self?.route(msg)
            }
        }
    }

    /// Simulate the daemon connecting over the transport.
    public nonisolated func connect(name: String = "Car Thing") {
        adapter.connect(Device(id: deviceId, name: name))
    }

    private func route(_ msg: GatewayToBridgeMsg) {
        if case let .response(meta) = msg.meta, let cont = pending.removeValue(forKey: meta.requestId) {
            cont.resume(returning: msg)
            return
        }
        if let idx = waiters.firstIndex(where: { $0.predicate(msg) }) {
            let waiter = waiters.remove(at: idx)
            waiter.cont.resume(returning: msg)
            return
        }
        outbound.append(msg)
    }

    private func failPending(_ id: UUID, _ error: Error) {
        if let cont = pending.removeValue(forKey: id) { cont.resume(throwing: error) }
    }

    private func failWaiter(_ id: UUID, _ error: Error) {
        if let idx = waiters.firstIndex(where: { $0.id == id }) {
            let waiter = waiters.remove(at: idx)
            waiter.cont.resume(throwing: error)
        }
    }

    /// Send a `.request` and await the matching `.response` frame, or throw on timeout.
    @discardableResult
    public func request(_ data: BridgeToGatewayMsgData, timeout: Duration = .seconds(5)) async throws -> GatewayToBridgeMsg {
        let id = UUID()
        let msg = BridgeToGatewayMsg(id: id, meta: .request, data: data)
        let frame = try codec.encode(msg)
        return try await withCheckedThrowingContinuation { cont in
            pending[id] = cont
            adapter.feed(deviceId: deviceId, frame)
            Task { [weak self] in
                try? await Task.sleep(for: timeout)
                await self?.failPending(id, WireDriverError.timeout)
            }
        }
    }

    /// Send a fire-and-forget command/event frame (no response expected).
    public func send(_ data: BridgeToGatewayMsgData, meta: MsgMeta = .command) async throws {
        let msg = BridgeToGatewayMsg(id: UUID(), meta: meta, data: data)
        adapter.feed(deviceId: deviceId, try codec.encode(msg))
    }

    /// Await the next outbound frame matching `predicate` (drains any already buffered first).
    public func waitOutbound(
        timeout: Duration = .seconds(5),
        where predicate: @escaping @Sendable (GatewayToBridgeMsg) -> Bool
    ) async throws -> GatewayToBridgeMsg {
        if let idx = outbound.firstIndex(where: predicate) {
            return outbound.remove(at: idx)
        }
        let id = UUID()
        return try await withCheckedThrowingContinuation { cont in
            waiters.append((id: id, predicate: predicate, cont: cont))
            Task { [weak self] in
                try? await Task.sleep(for: timeout)
                await self?.failWaiter(id, WireDriverError.timeout)
            }
        }
    }

    /// Snapshot of outbound frames seen so far (responses are consumed by `request`).
    public func outboundFrames() -> [GatewayToBridgeMsg] { outbound }

    public func stop() {
        pumpTask?.cancel()
        pumpTask = nil
    }
}

File: packages/companion/kotlin/companion/src/testShared/kotlin/com/bridgething/companion/WireDriver.kt
package com.bridgething.companion

import com.bridgething.gateway.AdapterEvent
import com.bridgething.gateway.Codec
import com.bridgething.gateway.Compression
import com.bridgething.gateway.Device
import com.bridgething.gateway.Encoding
import com.bridgething.schema.BridgeToGatewayMsg
import com.bridgething.schema.BridgeToGatewayMsgData
import com.bridgething.schema.GatewayToBridgeMsg
import com.bridgething.schema.MsgMeta
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

class WireDriver(
    private val adapter: FakeAdapter,
    val deviceId: String = "carthing-test",
    private val codec: Codec = Codec(defaultCompression = Compression.NONE, defaultEncoding = Encoding.MSGPACK),
) {
    private val pending = ConcurrentHashMap<UUID, CompletableDeferred<GatewayToBridgeMsg>>()
    private val outbound = Channel<GatewayToBridgeMsg>(Channel.UNLIMITED)
    private val buffered = ArrayDeque<GatewayToBridgeMsg>()
    private var pump: Job? = null

    fun start(scope: CoroutineScope) {
        pump = scope.launch {
            for ((_, frame) in adapter.sentFrames) {
                val msg = runCatching { codec.decode(GatewayToBridgeMsg.serializer(), frame) }.getOrNull() ?: continue
                val meta = msg.meta
                if (meta is MsgMeta.Response) {
                    pending.remove(meta.data.requestId)?.complete(msg)
                } else {
                    outbound.trySend(msg)
                }
            }
        }
    }

    suspend fun connect(name: String = "Car Thing") {
        adapter.simulate(AdapterEvent.Connected(Device(deviceId, name)))
    }

    suspend fun request(data: BridgeToGatewayMsgData, timeout: Duration = 5.seconds): GatewayToBridgeMsg {
        val id = UUID.randomUUID()
        val deferred = CompletableDeferred<GatewayToBridgeMsg>()
        pending[id] = deferred
        val frame = codec.encode(
            BridgeToGatewayMsg.serializer(),
            BridgeToGatewayMsg(id = id, meta = MsgMeta.Request, data = data),
        )
        adapter.simulate(AdapterEvent.Bytes(deviceId, frame))
        return withTimeout(timeout) { deferred.await() }
    }

    suspend fun send(data: BridgeToGatewayMsgData, meta: MsgMeta = MsgMeta.Command) {
        val frame = codec.encode(
            BridgeToGatewayMsg.serializer(),
            BridgeToGatewayMsg(id = UUID.randomUUID(), meta = meta, data = data),
        )
        adapter.simulate(AdapterEvent.Bytes(deviceId, frame))
    }

    suspend fun waitOutbound(
        timeout: Duration = 5.seconds,
        predicate: (GatewayToBridgeMsg) -> Boolean,
    ): GatewayToBridgeMsg = withTimeout(timeout) {
        val hit = buffered.indexOfFirst(predicate)
        if (hit >= 0) return@withTimeout buffered.removeAt(hit)
        var msg = outbound.receive()
        while (!predicate(msg)) {
            buffered.addLast(msg)
            msg = outbound.receive()
        }
        msg
    }

    fun stop() {
        pump?.cancel()
    }
}
